add consistent-trees HDF5 frontend

brittonsmith opened the pull request:
## PR Summary

This adds a new frontend for the HDF5 consistent-trees format implemented by @manodeep.

Using the sample data provided in Issue #46, one should be able to do:
```
import ytree
a = ytree.load('forest.h5', access='forest')
```
The access keyword can be set to either `tree` or `forest` to control whether trees or forests are loaded. In addition, one can now do:
```
forest = a[0]
print (a['forest']) # all nodes in the forest
print (a['tree']) # all nodes in the tree (you could always do this)
```
This new forest selection works with the `LHaloTree` frontend, which is already loading forests.

### Some current issues: *all solved*
- ~~some missing metadata could be helpful, if added to the data~~ - we now get these from the metadata attribute in the file
   - [x] cosmological parameters
   - [x] box size
   - [x] units for fields (perhaps as attributes on each field array dataset)
- ~~the `Input_Filename` filename values appear to not be written correctly~~ - these can be decoded as pointed out in this discussion and these are also not needed.

## PR Checklist

- [ ] Code passes tests.
- [ ] New features are documented with docstrings and narrative docs.
- [ ] Tests added for fixed bugs or new features.

brittonsmith commented:
BTW, I expect all the tests to fail since yt just merged it's `master` and `yt-4.0` branches. I have yet to update ytree to use `yt-4.0`, but I'll do that soon.

manodeep commented:
>    some missing metadata could be helpful, if added to the data
>>        cosmological parameters
>>        box size

Ooops - yes! These were part of the data-format design but I forgot to implement

>>        units for fields (perhaps as attributes on each field array dataset)

Ooh. I am not sure about this. Not everything has units, but I will try

>>    the Input_Filename filename values appear to not be written correctly

Ughhh, strings in hdf5 are such a pain. Is the error caused by null bytes in the filename? And does that go away if you individually decode each entry (i.e., once per tree)?

brittonsmith commented:
> 
> Ooh. I am not sure about this. Not everything has units, but I will try
> 

For fields without units, it would be fine to simply not write a `units` attribute for that dataset or to write an empty string. If it helps, here is the output from parsing the ctrees ascii header (which happens for the ascii ctrees frontend).

http://paste.yt-project.org/show/248/

> > > the Input_Filename filename values appear to not be written correctly
> 
> Ughhh, strings in hdf5 are such a pain. Is the error caused by null bytes in the filename? And does that go away if you individually decode each entry (i.e., once per tree)?

Just for reference, here is an example of what the strings look like when I load them with h5py. I don't think they can be rescued. That said, I don't actually need this information for anything.

http://paste.yt-project.org/show/247/

manodeep commented:
How do you automatically generate the units from the ascii file? I can simply add that bit of code into the ascii->hdf5 converter.

If you look carefully, you will see correct characters in that stream of binary (`/home/msinha/scratch/simulations...`) - that gives me hope. But clearly I need a better way to write those strings into the hdf5 file.

brittonsmith commented:
> How do you automatically generate the units from the ascii file? I can simply add that bit of code into the ascii->hdf5 converter.

This might not be useful to you since it's Python. This is ripped right out of `ytree`.
https://gist.github.com/brittonsmith/9ac6ce00d402bda888f95ee5bb6b7a17
You can also just hard code with this if you want. The output looks like this:
http://paste.yt-project.org/show/249/
 
> If you look carefully, you will see correct characters in that stream of binary (`/home/msinha/scratch/simulations...`) - that gives me hope. But clearly I need a better way to write those strings into the hdf5 file.

That's true, those ones went fine. Although, maybe the absolute path should be removed?

manodeep commented on ytree/frontends/consistent_trees_hdf5/arbor.py:
```diff
@@ -0,0 +1,113 @@
+            return
+
+        file_offsets = self._file_count.cumsum() - self._file_count
+        pbar = get_pbar('Planting trees', self._ntrees)
```
Should this be forests instead of trees? 

All the trees belonging to the same forest are written contiguously, and the relevant start and stop indices are written within the [ForestInfo dataset](https://gist.github.com/manodeep/1ce949bb51be1d93ba1ce374f556bc2d#file-convert_uchuu_ctrees_to_h5-py-L246-L253)

brittonsmith replied:
I guess I don't really know. We're loading with the `TreeInfo` dataset, but I have to admit I have never understood what a forest is. Now seems like a really good time to ask. What is a forest?

manodeep replied:
A tree has one FOF halo at the end of the simulation. Say a flyby occurs, a FofB falls into FofA, becomes a subhalo but then detaches and becomes a FofB again. By definition, FofA and FofB will be in two different trees, but in the snapshots where FofB was a subhalo, the hostID (upid) of FofB will be set to FofA. Then, to get a complete picture of the halo/galaxy evolution, you need to load both the FofA and FofB trees. A forest is simply this collection of trees such that **all** halo interactions are captured. A forest without flybys will only have one tree and the forests containing massive halos (i.e., clusters) have had lots of flybys and will consist of many trees. 

TLDR: Trees can contain references to halos in other trees, forests do not.

I can see both use-cases from a viz. perspective - loading by trees or loading by forests. From a semi-analytic model perspective, processing *must* occur in units of forests.

manodeep commented:
I looked into the `Input_Filename` field - I could not figure out how to incorporate a variable length string into the compound datatype for the `TreeInfo` file. 

The way to reconstruct the list of (string) input filenames:

```python
hf = h5py.File(<fname>, 'r')
input_filenames = [x.decode().replace('\x00', '') for x in hf['TreeInfo']['Input_Filename']]
```
Here's what the output looks like on our local linux cluster:

```python
In [92]: hf['TreeInfo']['Input_Filename'][0]
Out[92]: b'/\x00\x00\x00h\x00\x00\x00o\x00\x00\x00m\x00\x00\x00e\x00\x00\x00/\x00\x00\x00m\x00\x00\x00s\x00\x00\x00i\x00\x00\x00n\x00\x00\x00h\x00\x00\x00a\x00\x00\x00/\x00\x00\x00s\x00\x00\x00c\x00\x00\x00r\x00\x00\x00a\x00\x00\x00t\x00\x00\x00c\x00\x00\x00h\x00\x00\x00/\x00\x00\x00s\x00\x00\x00i\x00\x00\x00m\x00\x00\x00u\x00\x00\x00l\x00\x00\x00a\x00\x00\x00t\x00\x00\x00i\x00\x00\x00o\x00\x00\x00n\x00\x00\x00s\x00\x00\x00/\x00\x00\x00u\x00\x00\x00c\x00\x00\x00h\x00\x00\x00u\x00\x00\x00u\x00\x00\x00/\x00\x00\x00U\x00\x00\x002\x00\x00\x000\x00\x00\x000\x00\x00\x000\x00\x00\x00/\x00\x00\x00t\x00\x00\x00r\x00\x00\x00e\x00\x00\x00e\x00\x00\x00s\x00\x00\x00/\x00\x00\x000\x00\x00\x000\x00\x00\x000\x00\x00\x000\x00\x00\x000\x00\x00\x001\x00\x00\x004\x00\x00\x008\x00\x00\x00.\x00\x00\x00t\x00\x00\x00r\x00\x00\x00e\x00\x00\x00e'
In [93]: hf['TreeInfo']['Input_Filename'][0].decode().replace('\x00', '')
Out[93]: '/home/msinha/scratch/simulations/uchuu/U2000/trees/00000148.tree'
```

manodeep commented:
Since the halos can be written as either an array of structures or a structure of arrays, I am planning to add the units as a separate dataset

brittonsmith commented:
Thanks for your responses, Manodeep. I've been buried this week, but will get back on track next week. I think we might be able to support both trees and forests without much more effort. I'll get back to you with more details real soon.

brittonsmith commented:
@manodeep, I've started a `ytree` channel on the [yt project Slack](https://yt-project.org/slack.html). We can continue more detailed conversation there if it's easier.

brittonsmith commented:
> I looked into the `Input_Filename` field - I could not figure out how to incorporate a variable length string into the compound datatype for the `TreeInfo` file.
> 
> The way to reconstruct the list of (string) input filenames:
> 
> ```python
> hf = h5py.File(<fname>, 'r')
> input_filenames = [x.decode().replace('\x00', '') for x in hf['TreeInfo']['Input_Filename']]
> ```
> 
> Here's what the output looks like on our local linux cluster:
> 
> ```python
> In [92]: hf['TreeInfo']['Input_Filename'][0]
> Out[92]: b'/\x00\x00\x00h\x00\x00\x00o\x00\x00\x00m\x00\x00\x00e\x00\x00\x00/\x00\x00\x00m\x00\x00\x00s\x00\x00\x00i\x00\x00\x00n\x00\x00\x00h\x00\x00\x00a\x00\x00\x00/\x00\x00\x00s\x00\x00\x00c\x00\x00\x00r\x00\x00\x00a\x00\x00\x00t\x00\x00\x00c\x00\x00\x00h\x00\x00\x00/\x00\x00\x00s\x00\x00\x00i\x00\x00\x00m\x00\x00\x00u\x00\x00\x00l\x00\x00\x00a\x00\x00\x00t\x00\x00\x00i\x00\x00\x00o\x00\x00\x00n\x00\x00\x00s\x00\x00\x00/\x00\x00\x00u\x00\x00\x00c\x00\x00\x00h\x00\x00\x00u\x00\x00\x00u\x00\x00\x00/\x00\x00\x00U\x00\x00\x002\x00\x00\x000\x00\x00\x000\x00\x00\x000\x00\x00\x00/\x00\x00\x00t\x00\x00\x00r\x00\x00\x00e\x00\x00\x00e\x00\x00\x00s\x00\x00\x00/\x00\x00\x000\x00\x00\x000\x00\x00\x000\x00\x00\x000\x00\x00\x000\x00\x00\x001\x00\x00\x004\x00\x00\x008\x00\x00\x00.\x00\x00\x00t\x00\x00\x00r\x00\x00\x00e\x00\x00\x00e'
> In [93]: hf['TreeInfo']['Input_Filename'][0].decode().replace('\x00', '')
> Out[93]: '/home/msinha/scratch/simulations/uchuu/U2000/trees/00000148.tree'
> ```

Ok, this is fine. I don't think we'll ever need this field in ytree, but we can leave a note in the docs for anyone who might want to access it.

manodeep commented:
> > I looked into the `Input_Filename` field - I could not figure out how to incorporate a variable length string into the compound datatype for the `TreeInfo` file.
> > The way to reconstruct the list of (string) input filenames:
> > ```python
> > hf = h5py.File(<fname>, 'r')
> > input_filenames = [x.decode().replace('\x00', '') for x in hf['TreeInfo']['Input_Filename']]
> > ```
> > 
> > 
> > Here's what the output looks like on our local linux cluster:
> > ```python
> > In [92]: hf['TreeInfo']['Input_Filename'][0]
> > Out[92]: b'/\x00\x00\x00h\x00\x00\x00o\x00\x00\x00m\x00\x00\x00e\x00\x00\x00/\x00\x00\x00m\x00\x00\x00s\x00\x00\x00i\x00\x00\x00n\x00\x00\x00h\x00\x00\x00a\x00\x00\x00/\x00\x00\x00s\x00\x00\x00c\x00\x00\x00r\x00\x00\x00a\x00\x00\x00t\x00\x00\x00c\x00\x00\x00h\x00\x00\x00/\x00\x00\x00s\x00\x00\x00i\x00\x00\x00m\x00\x00\x00u\x00\x00\x00l\x00\x00\x00a\x00\x00\x00t\x00\x00\x00i\x00\x00\x00o\x00\x00\x00n\x00\x00\x00s\x00\x00\x00/\x00\x00\x00u\x00\x00\x00c\x00\x00\x00h\x00\x00\x00u\x00\x00\x00u\x00\x00\x00/\x00\x00\x00U\x00\x00\x002\x00\x00\x000\x00\x00\x000\x00\x00\x000\x00\x00\x00/\x00\x00\x00t\x00\x00\x00r\x00\x00\x00e\x00\x00\x00e\x00\x00\x00s\x00\x00\x00/\x00\x00\x000\x00\x00\x000\x00\x00\x000\x00\x00\x000\x00\x00\x000\x00\x00\x001\x00\x00\x004\x00\x00\x008\x00\x00\x00.\x00\x00\x00t\x00\x00\x00r\x00\x00\x00e\x00\x00\x00e'
> > In [93]: hf['TreeInfo']['Input_Filename'][0].decode().replace('\x00', '')
> > Out[93]: '/home/msinha/scratch/simulations/uchuu/U2000/trees/00000148.tree'
> > ```
> 
> Ok, this is fine. I don't think we'll ever need this field in ytree, but we can leave a note in the docs for anyone who might want to access it.

Yup you should not need this field within ytree. this field is there for tests and data provenance

manodeep commented on ytree/frontends/consistent_trees_hdf5/arbor.py:
```diff
@@ -30,6 +30,21 @@
 
+_access_names = {
+    'tree':   {'group'     : 'TreeInfo',
+               'parent_id' : 'TreeRootID',
```
Instead of `parent_id`, is it possible to use `unique_id`?

brittonsmith replied:
Absolutely, will change.

brittonsmith commented:
@manodeep, we now use the metadata attribute to scrape units and parameters. The frontend should now be fully functional. When you have time, go ahead and put it through its paces. When you're satisfied things look ok, I'll finish updating the documentation to mention the new frontend. We will also need to add some official sample so we can run automated testing. Would you be ok with using the data set you provided?

manodeep commented on ytree/frontends/consistent_trees/utilities.py:
```diff
@@ -0,0 +1,144 @@
+    # Read the first line as a list of all fields.
+    # Do some footwork to remove awkard characters.
+    rfl = next_line()[1:].strip().split()
+    reg = re.compile(r"\(\d+\)$")
```
You might need to remove more special characters to make the field names the same as the CTrees-hdf5 that I am producing. This regex leaves in the characters `[, ], /, |`. Here's what I have used:

```python
def _sanitize_ctrees_header(headerline):
    import re

    header = [re.sub('\(\d+\)$', '', s) for s in headerline]
    # print("After normal sub: header = {}\n".format(header))
    header = [re.sub('[^a-zA-Z0-9 \n\.]', '_', s) for s in header]
    # print("After removing special characters: header = {}\n".format(header))
    header = [re.sub('_$', '', s) for s in header]
    # print("After replacing trailing underscore: header = {}\n".format(header))
    header = [re.sub('(_)+', '_', s) for s in header]
    # print("After replacing multiple underscores: header = {}".format(header))
    return header
```

manodeep commented on ytree/frontends/consistent_trees_hdf5/arbor.py:
```diff
@@ -82,22 +95,38 @@ def _get_data_files(self):
 
-        self.field_list = list(fi.keys())
-        self.field_info.update(fi)
+        header_fi = parse_ctrees_header(
```
Ahha - I see. Here's where you have fixed the field names. 

Sorrry - should have told you :D

brittonsmith replied:
Yeah, I do it all down here. That said, I think it makes sense to run the same regex operations rather than reverse engineering a solution. I'll implement your suggestions above in place of what I've done.

brittonsmith replied:
I changed the string manipulation block to use regexes to be a little bit clearer. Hmm, that may be the first time in history anyone has said that.

manodeep commented on ytree/frontends/consistent_trees/utilities.py:
```diff
@@ -69,7 +69,7 @@ def next_line():
         if line is None:
             if lines_after_header:
```
What does the `lines_after_header` variable represent? (I don't have the context but those two lines  - `if line is None: if lines_after_header` look strange)

brittonsmith replied:
`lines_after_header` is a bad name for this variable. I couldn't think of anything better. This allows this function to handle three different use cases:
1. conventional ctrees files (`lines_after_header=True`) - the first line after the header is a single integer containing the number of trees. We want to read that and also record the location in the file where the actual data starts. Also, if there is nothing after the header, raise an exception because this file is incomplete. Even an empty tree should have a 0 there.
2. ctrees hlist files (`'lines_after_header=False`) - the headers are identical to case 1 except for the integer with the number of trees. That won't be there, so skip it.
3. ctrees hdf5 (`lines_after_header=False`) - like case 2, just parse the lines from the header itself and quit.

I'm open to suggestions for new names. I can also restructure this block if you think it would be beneficial to someone reading this.

manodeep replied:
How about `has_nforests_in_file`?

brittonsmith replied:
I went with `ntrees_in_file`. Not quite as long, reasonably accurate.

manodeep replied:
Yup - that works

brittonsmith commented:
@manodeep, if everything looks ok now, I'll update the documentation to discuss this frontend. Are there any specific information or links you'd like me to include when discussing this?

manodeep commented:
I am attempting to use the new functionality - will report back in the next few days.

Relevant links would be the [uchuutools package](https://github.com/uchuuproject/uchuutools), and the Uchuu simulations DR1 under [here](http://www.skiesanduniverses.org/). Both those links should contain significantly more code/data in the next week or so.

manodeep commented:
Hit an error when executing `ytree.load()`

```python
In [3]: a = ytree.load('/home/msinha/scratch/simulations/uchuu/U2000/trees/hdf5/with-tree-indices/forest.h5')
Performance will be improved by saving this arbor with "save_arbor" and reloading:
        >>> a = ytree.load("/home/msinha/scratch/simulations/uchuu/U2000/trees/hdf5/with-tree-indices/forest.h5")
        >>> fn = a.save_arbor()
        >>> a = ytree.load(fn)
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
<ipython-input-3-abca60fc0602> in <module>()
----> 1 a = ytree.load('/home/msinha/scratch/simulations/uchuu/U2000/trees/hdf5/with-tree-indices/forest.h5')

/home/msinha/research/codes/ytree/ytree/data_structures/arbor.py in load(filename, method, **kwargs)
   1151              "\t>>> a = ytree.load(fn)") % filename)
   1152         load_warn = False
-> 1153     return arbor_registry[method](filename, **kwargs)

/home/msinha/research/codes/ytree/ytree/frontends/consistent_trees_hdf5/arbor.py in __init__(self, filename, access)
     67                  'Valid options are: %s.') % (access, _access_names))
     68         self.access = access
---> 69         super(ConsistentTreesHDF5Arbor, self).__init__(filename)
     70
     71     def _node_io_loop_prepare(self, root_nodes):

/home/msinha/research/codes/ytree/ytree/data_structures/arbor.py in __init__(self, filename)
     97         self.directory = os.path.dirname(filename)
     98         self._parse_parameter_file()
---> 99         self._set_units()
    100         self._setup_io()
    101         self._get_data_files()

/home/msinha/research/codes/ytree/ytree/data_structures/arbor.py in _set_units(self)
    126             omega_matter=self.omega_matter,
    127             omega_lambda=self.omega_lambda,
--> 128             unit_registry=self.unit_registry)
    129
    130     def _setup_io(self):

/home/msinha/.local/lib/python3.6/site-packages/yt/utilities/cosmology.py in __init__(self, hubble_constant, omega_matter, omega_lambda, omega_radiation, omega_curvature, unit_registry, unit_system, use_dark_factor, w_0, w_a)
    106                                   dimensions.length, "\\rm{%s}/(1+z)" % my_unit)
    107         self.unit_registry = unit_registry
--> 108         self.hubble_constant = self.quan(hubble_constant, "100*km/s/Mpc")
    109         self.unit_system = unit_system
    110

/home/msinha/.local/lib/python3.6/site-packages/yt/units/yt_array.py in __new__(cls, input_scalar, input_units, registry, dtype, bypass_validation)
   1599             raise RuntimeError("YTQuantity values must be numeric")
   1600         ret = YTArray.__new__(cls, input_scalar, input_units, registry,
-> 1601                               dtype=dtype, bypass_validation=bypass_validation)
   1602         if ret.size > 1:
   1603             raise RuntimeError("YTQuantity instances must be scalars")

/home/msinha/.local/lib/python3.6/site-packages/yt/units/yt_array.py in __new__(cls, input_array, input_units, registry, dtype, bypass_validation)
    522             # don't handle all the cases here, let the Unit class handle if
    523             # it's a str.
--> 524             units = Unit(input_units, registry=registry)
    525
    526         # Attach the units

/home/msinha/.local/lib/python3.6/site-packages/yt/units/unit_object.py in __new__(cls, unit_expr, base_value, base_offset, dimensions, registry, latex_repr, **assumptions)
    199                 unit_expr = unit_expr.decode("utf-8")
    200
--> 201             if registry and unit_expr in registry.unit_objs:
    202                 return registry.unit_objs[unit_expr]
    203             else:

AttributeError: 'UnitRegistry' object has no attribute 'unit_objs'

In [4]: 
```

Not that it should matter, but I installed `ytree` with `python -m pip install --user -e .` after a `git clone <brittonsmith/ytree> && git checkout h5ctrees`. Here are the modules loaded on the cluster that I was working on
```sh
[~ @farnarkle1] ml

Currently Loaded Modules:
  1) slurm/.latest  (H,S)   5) gcc/6.4.0         9) scalapack/2.0.2-openblas-0.2.20  13) numpy/1.16.3-python-3.6.4  17) openpgm/5.2.122                 21) git/2.18.0
  2) nvidia/.latest (H,S)   6) openmpi/3.0.0    10) python/3.6.4                     14) sqlite/3.21.0              18) zeromq/4.2.3
  3) binutils/2.30          7) openblas/0.2.20  11) szip/2.1.1                       15) hdf5/1.10.1                19) pyzmq/16.0.4-python-3.6.4-zmq4
  4) gcccore/6.4.0          8) fftw/3.3.7       12) mpi4py/3.0.0-python-3.6.4        16) h5py/2.7.1-python-3.6.4    20) ipython/5.5.0-python-3.6.4

  Where:
   S:  Module is Sticky, requires --force to unload or purge
   H:             Hidden Module

 

```

brittonsmith commented:
@manodeep, I made the questionable choice of making the ytree development branch depend on yt's development branch, which is now pre-release for yt-4.0. This means you'll have to install the yt `master` branch by cloning it's repo from `github.com/yt-project/yt` and doing a similar pip install as you tried above. You may also need to update your install of `unyt` to at least version 2.7.2. Hopefully those two things are all you need.

manodeep commented:
Thanks. I can get ytree to load the forests with both 'tree' and 'forest' access. I followed the instructions on [here](https://ytree.readthedocs.io/en/latest/Plotting.html#making-tree-plots) and could make a plot of the mergertree. 

How do I plot the biggest forest (by the number of halos) across all the files? Or better yet, how do I plot the forest that contains the most massive FOF halo at z=0?

manodeep commented:
And given how long it takes to load up the mass field across all files, can ytree be run in parallel (either mpi or multiprocesses)?

brittonsmith commented:
The easiest way to get the most massive halo (or the one with the most nodes) would be something like this:
```
a = ytree.load(...)
imax = a['mass'].argmax()
# or this for nodes
# tree_size = a.arr([t.tree_size for t in a])
# imax = tree_size.argmax()
my_tree = a[imax]
```

To save time, you could then save just that tree (or forest) to a separate data set.
```
fn = my_tree.save_tree()
a = ytree.load(fn)
```

Currently, there's no way to specify a subset of data at load time. How would you like to see that work? An additional keyword arg to `load` specifying certain files?

As for parallelism, it's not there yet, but it's coming. I'm working on something that will use dask, but I'm learning it will require a small amount of refactoring. Given that this it's on it's way, we should try to get this merged first to avoid conflicts. I propose we narrow down anything remaining to make this base-level functional and then merge after that. We can always issue more PRs with updates and supporting the `aos` format can also come a little later. What do you think?

manodeep commented:
Regarding parallel capabilities, 100% agree with you. Let's merge this in first before mucking about in parallel capabilities. 

Just so I understand correctly when you do `imax = a['mass'].argmax()`, you are first having to load the masses of *all* halos across all trees right? 

To specifically load a tree, it would have to be a keyword - ideally, something like a `treeID` but not all tree-formats carry such a field. Treenumber would be the other choice but that would either be ambiguous (i.e. file-local vs global-across-all-files), require additional work on the users part (e.g., convert from file-local to global). For instance, ascii-consistent-trees, it would not even make sense to provide access through tree-number. Do you think it would be feasible to provide both kinds of access - through TreeID/ForestID value (and matching on that specific column), *and* through tree-number + keyword to disambiguate file-local/global ?

Somewhat related: Right now the `is_valid` function for ctrees-hdf5 is checking whether the file is the container `forest.h5`. Is the expectation that `ytree` always has to be invoked with *all* the forests? For instance, what if the user only has downloaded the `forest_0.h5` and wants to explore the trees/forests in that one file?

brittonsmith commented:
> Just so I understand correctly when you do `imax = a['mass'].argmax()`, you are first having to load the masses of _all_ halos across all trees right?

I've just merged this with a refactor effort I completed this week that should speed things up quite a bit. I'm seeing about a factor 50 speedup in the planting trees/forests step and about a factor 20 speedup in reading a single field for the roots of all trees/forests, i.e., the thing that happens when you do `a['mass']`. We loop through each file group, read the full array for that field, and then numpy slice it to get the first value for each tree/forest. After this loop, we do a concatenate of the slices arrays. The class I implemented in the most recent changesets should give a ~20x speedup from what you would have been seeing. It should also parallelize quite easily.

> To specifically load a tree, it would have to be a keyword - ideally, something like a `treeID` but not all tree-formats carry such a field. Treenumber would be the other choice but that would either be ambiguous (i.e. file-local vs global-across-all-files), require additional work on the users part (e.g., convert from file-local to global). For instance, ascii-consistent-trees, it would not even make sense to provide access through tree-number. Do you think it would be feasible to provide both kinds of access - through TreeID/ForestID value (and matching on that specific column), _and_ through tree-number + keyword to disambiguate file-local/global ?

If the user knows only the ID, then doing the following would be the most straightforward:
```
a = ytree.load(...)
my_tree = a[a['uid'] == ID]
# or technically this could work and be faster (i can connect this to the above as well)
a._plant_trees()
my_tree = a[a._node_info['uid'] == ID]
```
If they knew what file(s) it was in, we could implement something like:
```
a = ytree.load(..., input_files=[...])
```
Then, we would only make available trees/forests that live in those files and skip everything else. That would make the initial pass of planting all the trees a little bit faster. What do you think of that?

> Somewhat related: Right now the `is_valid` function for ctrees-hdf5 is checking whether the file is the container `forest.h5`. Is the expectation that `ytree` always has to be invoked with _all_ the forests? For instance, what if the user only has downloaded the `forest_0.h5` and wants to explore the trees/forests in that one file?

My understanding is that if the data set is comprised of `forest.h5` and `forest_0...N.h5` files, then if the user just wanted one of the data files, they would always have to have `forest.h5` as well since that has the metadata and virtual datasets. Is that how you see things working for using a subset of the data?

manodeep commented:
> > Just so I understand correctly when you do `imax = a['mass'].argmax()`, you are first having to load the masses of _all_ halos across all trees right?
> 
> I've just merged this with a refactor effort I completed this week that should speed things up quite a bit. I'm seeing about a factor 50 speedup in the planting trees/forests step and about a factor 20 speedup in reading a single field for the roots of all trees/forests, i.e., the thing that happens when you do `a['mass']`. We loop through each file group, read the full array for that field, and then numpy slice it to get the first value for each tree/forest. After this loop, we do a concatenate of the sliced arrays. The class I implemented in the most recent changesets should give a ~20x speedup from what you would have been seeing. It should also parallelize quite easily.
> 

Hmm - in the case where access is via 'forest', what does "root" mean? For Consistent-Trees, since the forest is comprised of multiple trees, and I am not certain, what/how defines the 'forest' other than the collection. For my use-case, I would want to know the maximum of halo mass across all halos at the final snapshot contained in all trees (or a single tree if access is 'trees'). 


> > To specifically load a tree, it would have to be a keyword - ideally, something like a `treeID` but not all tree-formats carry such a field. Treenumber would be the other choice but that would either be ambiguous (i.e. file-local vs global-across-all-files), require additional work on the users part (e.g., convert from file-local to global). For instance, ascii-consistent-trees, it would not even make sense to provide access through tree-number. Do you think it would be feasible to provide both kinds of access - through TreeID/ForestID value (and matching on that specific column), _and_ through tree-number + keyword to disambiguate file-local/global ?
> 
> If the user knows only the ID, then doing the following would be the most straightforward:
> 
> ```
> a = ytree.load(...)
> my_tree = a[a['uid'] == ID]
> # or technically this could work and be faster (i can connect this to the above as well)
> a._plant_trees()
> my_tree = a[a._node_info['uid'] == ID]
> ```
> 
> If they knew what file(s) it was in, we could implement something like:
> 
> ```
> a = ytree.load(..., input_files=[...])
> ```
> 
> Then, we would only make available trees/forests that live in those files and skip everything else. That would make the initial pass of planting all the trees a little bit faster. What do you think of that?
> 

Thanks! That sounds great. Would this be generalizable to other supported tree-formats in ytree?

> > Somewhat related: Right now the `is_valid` function for ctrees-hdf5 is checking whether the file is the container `forest.h5`. Is the expectation that `ytree` always has to be invoked with _all_ the forests? For instance, what if the user only has downloaded the `forest_0.h5` and wants to explore the trees/forests in that one file?
> 
> My understanding is that if the data set is comprised of `forest.h5` and `forest_0...N.h5` files, then if the user just wanted one of the data files, they would always have to have `forest.h5` as well since that has the metadata and virtual datasets. Is that how you see things working for using a subset of the data?

The actual data live within `forest_0..N.h5` - so ytree should be able to load those files directly. Particularly, if there is an option to specify which `input_files` were wanted. I would say to change the `is_valid` function to check for `TotNforests/TotNtrees/TotNhalos` as well as their equivalent in single files `Nforests/Ntrees/Nhalos`. For instance, if the user only passed `forest.h5` then the expectation is all files under `File[0..N]` within `forest.h5` are to be loaded. If the user passes `[forest_(0..N).h5]`, then only those files are to be loaded. Does that make sense?

brittonsmith commented:
> > > Just so I understand correctly when you do `imax = a['mass'].argmax()`, you are first having to load the masses of _all_ halos across all trees right?
> > 
> > 
> > I've just merged this with a refactor effort I completed this week that should speed things up quite a bit. I'm seeing about a factor 50 speedup in the planting trees/forests step and about a factor 20 speedup in reading a single field for the roots of all trees/forests, i.e., the thing that happens when you do `a['mass']`. We loop through each file group, read the full array for that field, and then numpy slice it to get the first value for each tree/forest. After this loop, we do a concatenate of the sliced arrays. The class I implemented in the most recent changesets should give a ~20x speedup from what you would have been seeing. It should also parallelize quite easily.
> 
> Hmm - in the case where access is via 'forest', what does "root" mean? For Consistent-Trees, since the forest is comprised of multiple trees, and I am not certain, what/how defines the 'forest' other than the collection. For my use-case, I would want to know the maximum of halo mass across all halos at the final snapshot contained in all trees (or a single tree if access is 'trees').
> 

In that case, I think you would want to load with `access='trees'` and then do `a['mass'].argmax()` to get the most massive halo. I think you are right about the meaning of the forest id. As far as I can tell, it is just the unique id of the root of one of trees in that forest, but not necessarily the most massive or anything. If you are loading with `access='forest'`, then you can get the root nodes of a forest with something like the following:
```
# first forest
my_forest = a[0]
is_root = my_forest['forest', 'desc_uid'] == -1
roots = np.array(list(my_forest['forest']))[is_root]
```

> > > To specifically load a tree, it would have to be a keyword - ideally, something like a `treeID` but not all tree-formats carry such a field. Treenumber would be the other choice but that would either be ambiguous (i.e. file-local vs global-across-all-files), require additional work on the users part (e.g., convert from file-local to global). For instance, ascii-consistent-trees, it would not even make sense to provide access through tree-number. Do you think it would be feasible to provide both kinds of access - through TreeID/ForestID value (and matching on that specific column), _and_ through tree-number + keyword to disambiguate file-local/global ?
> > 
> > 
> > If the user knows only the ID, then doing the following would be the most straightforward:
> > ```
> > a = ytree.load(...)
> > my_tree = a[a['uid'] == ID]
> > # or technically this could work and be faster (i can connect this to the above as well)
> > a._plant_trees()
> > my_tree = a[a._node_info['uid'] == ID]
> > ```
> > 
> > 
> > If they knew what file(s) it was in, we could implement something like:
> > ```
> > a = ytree.load(..., input_files=[...])
> > ```
> > 
> > 
> > Then, we would only make available trees/forests that live in those files and skip everything else. That would make the initial pass of planting all the trees a little bit faster. What do you think of that?
> 
> Thanks! That sounds great. Would this be generalizable to other supported tree-formats in ytree?
> 
> > > Somewhat related: Right now the `is_valid` function for ctrees-hdf5 is checking whether the file is the container `forest.h5`. Is the expectation that `ytree` always has to be invoked with _all_ the forests? For instance, what if the user only has downloaded the `forest_0.h5` and wants to explore the trees/forests in that one file?
> > 
> > 
> > My understanding is that if the data set is comprised of `forest.h5` and `forest_0...N.h5` files, then if the user just wanted one of the data files, they would always have to have `forest.h5` as well since that has the metadata and virtual datasets. Is that how you see things working for using a subset of the data?
> 
> The actual data live within `forest_0..N.h5` - so ytree should be able to load those files directly. Particularly, if there is an option to specify which `input_files` were wanted. I would say to change the `is_valid` function to check for `TotNforests/TotNtrees/TotNhalos` as well as their equivalent in single files `Nforests/Ntrees/Nhalos`. For instance, if the user only passed `forest.h5` then the expectation is all files under `File[0..N]` within `forest.h5` are to be loaded. If the user passes `[forest_(0..N).h5]`, then only those files are to be loaded. Does that make sense?

Ok, I've implemented the following options:
```
# whole enchilada
a = ytree.load('forest.h5')
# single file
a = ytree.load('forest_0.h5')
# multiple files
a = ytree.load(['forest_0.h5', 'forest_1.h5'])
```
Those will all give the same behavior. I can see this being abstracted to the `LHaloTree` frontend at some point. Unfortunately, I'm already working on borrowed time, but I can put a pin in this and come back to it later.

manodeep commented:
All of that is amazing - thanks!

manodeep commented on ytree/data_structures/arbor.py:
```diff
@@ -110,16 +110,28 @@ def __init__(self, filename):
+        """
+
+        self.filename = filename
+        if isinstance(filename, list):
```
Is it worth including `tuple` in this? i.e., `if isinstance(filename, (list, tuple))` or does the documentation specify a list is required?

brittonsmith replied:
Sure, I can do tuple. I haven't added the documentation yet, so it's fine to add that.

manodeep commented on ytree/frontends/consistent_trees_hdf5/arbor.py:
```diff
@@ -0,0 +1,216 @@
+        aname = _access_names[self.access]['file_size']
+
+        if self._virtual_dataset:
+            with h5py.File(self.filename, mode='r') as f:
```
`self.filename` can be a `list` right? Does `h5py` return a list of open file handles in that case?

manodeep replied:
Or is that because of the virtual dataset that self.filename must be `forest.h5` and not a list of filenames

brittonsmith replied:
Yeah, that's right. `self.filename` will be just the one file. I used `self.filename` here instead of `self.parameter_filename` just to use the same variable through `_get_data_files`, but it would make no difference.

manodeep commented on ytree/frontends/consistent_trees_hdf5/arbor.py:
```diff
@@ -0,0 +1,216 @@
+
+    @property
+    def _virtual_dataset(self):
+        return re.search(r"\_\d+\.h5$", self.parameter_filename) is None
```
Is the `parameter_filename` set by the implementation? From the looks of it, if `forest.h5` is passed, then a `virtual_dataset` is being set

brittonsmith replied:
`parameter_filename` is set in the base class. It's just `self.filename` if `filename` is a string or the first item if it's a list.

brittonsmith commented:
I've completed docs and testing for this frontend. A rendered version of the updated docs can be found [here](https://ytree-fork.readthedocs.io/en/h5ctrees/). I think this PR is ready to be merged, pending any final comments.

Manodeep, if you're happy with this, go ahead and approve. After that, I propose we merge this, close Issue #46, and open a new issue to add support for the array-of-structs version. I can probably get to this sometime in the fall.

manodeep commented:
Looks great! Thanks **a lot** for implementing this. 

Agreed about incorporating AOS in the future. 

I will say that you should fix the comment typo though - I am fairly certain it will bug you. (`multitple` instead of `multiple`)

brittonsmith commented:
You're very welcome. I'm glad this could come together.

Thanks for spotting the typo; fixed.

I've gone ahead and added functional support for array of structs format. Loading data in that format will produce a warning that field access is much slower than struct of arrays format. We can think about how to better optimize for this format in the future.

Take one last look to make sure I haven't said anything incorrect in the updated docs. Also, Manodeep, I'd like to add you to the list of contributors for the next release. Let me know what you think!

manodeep commented:
Awesome - yup that should work for the `aos` case. Good idea about the warning too!

I didn't really contribute any code - are you sure you want to add me as a contributor? (I am happy to be added but don't want to dilute the efforts that you and Meagan put in)

brittonsmith commented:
Great, I will now merge this.

Maybe not code, but certainly plenty of feedback and guidance on this frontend and on helping me understand what to do with forests. Valuable contributions all. Anyway, you don't have to decide now. The next release is probably a little ways off.
